src/ucp/tag/rndv.c

```diff
@@ -596,6 +599,236 @@ static void ucp_rndv_req_send_rma_get(ucp_request_t *rndv_req, ucp_request_t *rr
+        }
+    }
+
+    req->recv.tag.remaining -= freq->send.length;
+    if (req->recv.tag.remaining == 0) {
+        ucp_request_complete_tag_recv(req, UCS_OK);
+    }
+
+    ucp_request_put(freq);
+}
+
+static void
+ucp_rndv_recv_frag_put_mem_type(ucp_request_t *rreq, ucp_request_t *rndv_req,
+                                ucp_request_t *freq, ucp_mem_desc_t *mdesc,
+                                size_t length, size_t offset)
+{
+    ucp_worker_h worker = rreq->recv.worker;
+    ucp_lane_index_t mem_type_rma_lane;
+    ucp_md_index_t md_index;
+    ucp_ep_h mem_type_ep;
+
+    ucs_assert_always(!UCP_MEM_IS_ACCESSIBLE_FROM_CPU(rreq->recv.mem_type));
+
+    /* PUT on memtype endpoint to stage from
+     * frag recv buffer to memtype recv buffer
+    */
+    mem_type_ep       = worker->mem_type_ep[rreq->recv.mem_type];
+    mem_type_rma_lane = ucp_ep_config(mem_type_ep)->key.rma_bw_lanes[0];
+
+    if (mem_type_rma_lane == UCP_NULL_LANE) {
+        ucs_fatal("no rma bw lane to stage from stage buffer to"
+                " memory type recv buffer");
+    }
+
+    md_index  = ucp_ep_md_index(mem_type_ep, mem_type_rma_lane);
+
+    ucp_request_send_state_init(freq, ucp_dt_make_contig(1), 0);
+    ucp_request_send_state_reset(freq, ucp_rndv_recv_frag_put_completion,
+                                 UCP_REQUEST_SEND_PROTO_RNDV_PUT);
+    freq->send.ep                         = mem_type_ep;
+    freq->send.lane                       = mem_type_rma_lane;
+    freq->send.buffer                     = mdesc + 1;
+    freq->send.datatype                   = ucp_dt_make_contig(1);
+    freq->send.mem_type                   = rreq->recv.mem_type;
+    freq->send.state.dt.dt.contig.memh[0] = ucp_memh2uct(mdesc->memh, md_index);
+    freq->send.state.dt.dt.contig.md_map  = UCS_BIT(md_index);
+    freq->send.length                     = length;
+    freq->send.uct.func                   = ucp_rndv_progress_rma_put_zcopy;
+    freq->send.rndv_put.sreq              = rreq;
+    freq->send.rndv_put.rkey              = NULL;
+    freq->send.rndv_put.remote_request    = (uintptr_t) rndv_req;
+    freq->send.rndv_put.remote_address    =
+                    (uint64_t)UCS_PTR_BYTE_OFFSET(rreq->recv.buffer, offset);
+    freq->send.mdesc                      = mdesc;
+
+    ucp_request_send(freq, 0);
+}
+
+static ucs_status_t
+ucp_rndv_send_frag_get_mem_type(ucp_request_t *sreq, uintptr_t rreq_ptr,
+                                size_t length, size_t offset,
+                                uct_completion_callback_t comp_cb)
+{
+    ucp_worker_h worker = sreq->send.ep->worker;
+    ucp_lane_index_t mem_type_rma_lane;
+    ucp_request_t *freq;
+    ucp_ep_h mem_type_ep;
+    ucp_mem_desc_t *mdesc;
+    ucp_md_index_t md_index;
+
+    /* GUT on memtype endpoint to stage from
+     * memtype send buffer to host frag buffer
+     */
+    freq = ucp_request_get(worker);
+    if (freq == NULL) {
+        ucs_fatal("failed to allocate fragment receive request");
```

Comment:
maybe also return `UCS_ERR_NO_MEMORY` here to be consistent with other checks below